### DISCUSSION TITLE
Pocketbase SDK with SvelteKit streaming issues

### DISCUSSION THREAD

**Discussion by selfagency** [pocketbase/pocketbase#5152] [Q&A] [answered]
What I'm trying to achieve is to query the full list from my Items collection, wait til it's loaded, and pass the results on to my Shop component. The problem I'm encountering is that the Pocketbase SDK's batch operations are still streaming requests in the background asynchronously after returning an initial value and SvelteKit's reactivity isn't picking up on the streamed changes to the value after the page's initial render. So even after `data.items` updates, the result of `items` remains undefined.

`+page.server.ts` where the async function is passed as the return value.

```ts
// types
import type { Item } from '$lib/types';

// internal
import { handleError, retry } from '$lib/server/api';

export async function load({ locals }) {
  const { api } = locals;

  const fetchItems = async (): Promise<Item[]> =>
    (await retry(
      async () =>
        await api.collection('items').getFullList({
          batch: 1000,
          expand: 'price'
        })
    )) as Item[];

  try {
    return {
      items: fetchItems(),
    };
  } catch (err) {
    return handleError(err as Error);
  }
}
```

`+page.svelte` where the async function is awaited and its value passed on.

```svelte
<script lang="ts">
  // internal components
  import Title from '$lib/components/global/Title.svelte';
  import Shop from '$lib/components/shop/Shop.svelte';
  import Error from '$lib/components/global/Error.svelte';

  // props
  export let data;
</script>

<svelte:head>
  <title>Supplies &middot; CNYRIC Supplies</title>
</svelte:head>

<section class="justify-start">
  <Title>Supplies</Title>

  <div class="w-full">
    {#await data.items}
      <Shop loading={true} />
    {:then items}
      {items.length}
      <Shop loading={false} {items} />
    {:catch error}
      <Error status={error.status} message={error.message} {data} />
    {/await}
  </div>
</section>
```

`Shop.svelte` which receives the value.

```svelte
<script lang="ts">

// [...]

  $: if (items.length > 0 || searchState) {
    log.debug('Items:', items.length);
    if (browser && worker) worker.postMessage({ data: items, searchState, year: $state.year.id });
  }
</script>

{items.length}
```

This is the output of the above log statement, which only ever triggers once.

```log
DEBUG	/src/lib/components/shop/Shop.svelte:141	Items: undefined
```

The component itself will eventually update to display the expected value of `{items.length}`, but the reactive statement itself never detects a change to the value, therefore never triggering my inital search filters.

Am I missing something/doing something wrong here?

Is there a way using the SDK to _not_ stream results from Pocketbase but fetch them all before returning a value, or do I have to write all my own fetch requests directly instead?

On a related note, it would be nice to be able to pass SvelteKit's custom `fetch` function (which allows for better SSR functionality) to the Pocketbase SDK, which accepts a custom `fetch` as a config option. However when you do so, for the very same reason I cite above, your logs will fill up with warnings:

```log
Calling `event.fetch(...)` in a promise handler after `load(...)` has returned will not cause the function to re-run when the dependency is invalidated
```

This is again because the `load` function is completing while the Pocketbase SDK is still processing async requests in the background that it will eventually update the returned data variable with.

**Comment by ganigeorgiev** [comment, marked as answer]
> The problem I'm encountering is that the Pocketbase SDK's batch operations are still streaming requests in the background asynchronously after returning an initial value.

Are you sure that this is not caused by the `retry` call? Isn't it required for the streaming to work the actual function to return a Promise and not the awaited final result as you seems to do with `await retry` and the `await api.collection`?

I don't exactly understand what is the issue you are experiencing and why there is need for so many callbacks, but the SDK doesn't do any streaming. `getFullList` calls `getList` on batches until all items are fetched. The returned value of the method is effectively single Promise that will be resolved once all operations are completed.

> On a related note, it would be nice to be able to pass SvelteKit's custom fetch function (which allows for better SSR functionality)...

You can already do that since [JS SDK v0.17+](https://github.com/pocketbase/js-sdk/blob/master/CHANGELOG.md). For more details please check https://github.com/pocketbase/js-sdk?tab=readme-ov-file#custom-request-options.

**Reply by ganigeorgiev** [reply]
Based on my understanding of the [SvelteKit docs](https://kit.svelte.dev/docs/load#streaming-with-promises), it should work if you replace your `fetchItems` function with directly the `getFullList` call, aka:

```js
return {
  items: api.collection('items').getFullList({
    batch: 500, // 500 is the maximum at the moment
    expand: 'price'
  })
};
```

**Reply by selfagency** [reply]
Dangit, it _was_ my `retry` function after all. Is there a way to retry requests built-in to the SDK?

**Reply by ganigeorgiev** [reply]
There is no builtin support for doing retries in SDK but I'm also not sure that I understand how this `retry` should work. 

Does it retry when the request fails? Does it have any rules, like wait time, max attempts, etc.?

~In general it should be doable~, but again in order for the streaming to work you need to return a `Promise` and not the awaited/resolved result.

_Edit: actually in case of the streaming I'm not sure how the retry will work as it is possible to fail on the 3rd batch request for example. You can overwrite/wrap the `fetch` request method and perform the retry from there, to repeat until it resolve to 20x response or some other condition._

**Reply by selfagency** [reply]
The issue is that if there are no results for a subdata type of a page I'm loading, but I want the page to still display with no results for that subdata type, the SDK will throw a 404 for the missing subdata and then the whole page will 404, as opposed to just returning the results of that missing subdata as empty.

I was handling that by specifying whether the retry function should throw or not, specifying `true` for the main datatype and leaving it `false` for the subtypes.

```
const retry = async (fn: () => Promise<unknown>, throwError: boolean = false): Promise<unknown> => {
  let retries = 3;
  let result: unknown | undefined = undefined;

  while (retries > 0) {
    try {
      retries--;
      result = await fn();
      if (result) {
        return result;
      }
    } catch (err) {
      log.error(err);
      if (retries === 0 && throwError) {
        throw err;
      } else {
        continue;
      }
    }
  }
};
```

**Reply by ganigeorgiev** [reply]
> The issue is that if there are no results for a subdata type of a page I'm loading, but I want the page to still display with no results for that subdata type, the SDK will throw a 404 for the missing subdata and then the whole page will 404, as opposed to just returning the results of that missing subdata as empty.

I've read it twice but I still don't understand what is the desired behavior. Note that `getFullList` shouldn't throw 404 and it is expected to return empty `[]` if it doesn't have any items (due to the List API rule or just because the collection is empty).

In any case, as mentioned in the SveteKit docs in order for the streaming to work you need to return the request's Promise and not `await`-ing it.

If you want to perform retries for the `getFullList` calls you'll have to perform it on `fetch` level because `getFullList` sends chained/consequent `getList` requests which will have to be retried individually in case they fail, aka. you'll have to do something like (pseudo-code):

```js
api.collection('items').getFullList({
  // batch: 500, the max batch size at the moment is always capped at 500
  expand: 'price',
  fetch: async (url, config) => {
    // wrap the original fetch in your custom retry code
    return retry(fetch(url, config))
  },
})
```

**Reply by selfagency** [reply]
I appreciate the guidance. I'm not always using `getFullList`. I also use `getFirstInList` and `getOne` based on ID numbers and basic filters.

FWIW, it apparently doesn't like `Promise.all` either.

```ts
// third-party dependencies
import cookie from 'cookie';

// internal
import { handleError } from '$lib/server/api';

export async function load({ locals, cookies, fetch }) {
  const { api } = locals;
  const user = JSON.parse(cookie.parse(cookies.get('auth') ?? '').pb_auth).model?.id;

  try {
    return {
      orderData: Promise.all([
        api.collection('orders').getFullList({
          filter: `placedBy='${user}'`,
          expand: 'district,placedBy,destination,budget,year',
          fetch,
        }),
        api.collection('orderItems').getFullList({
          filter: `order.placedBy='${user}'`,
          expand: 'item,price',
          fetch,
        }),
        api.collection('orderMeta').getFullList({ filter: `order.placedBy='${user}'`, fetch }),
      ]),
    };
  } catch (err) {
    return handleError(err as Error);
  }
}
```

```log
web    | src/routes/orders/+page.server.ts: Calling `event.fetch(...)` in a promise handler after `load(...)` has returned will not cause the function to re-run when the dependency is invalidated
web    | src/routes/orders/+page.server.ts: Calling `event.fetch(...)` in a promise handler after `load(...)` has returned will not cause the function to re-run when the dependency is invalidated
web    | src/routes/orders/+page.server.ts: Calling `event.fetch(...)` in a promise handler after `load(...)` has returned will not cause the function to re-run when the dependency is invalidated
web    | src/routes/orders/+page.server.ts: Calling `event.fetch(...)` in a promise handler after `load(...)` has returned will not cause the function to re-run when the dependency is invalidated
web    | src/routes/orders/+page.server.ts: Calling `event.fetch(...)` in a promise handler after `load(...)` has returned will not cause the function to re-run when the dependency is invalidated
```

**Comment by orenaksakal** [comment]
Streaming works well for me, this is how I do it with svelte 5 (works with svelte 4 too)


**the load function on +page.server.ts**
```
export const load = async (event) => {
	return {
			streaming: {
				renders: event.locals.pocketbase.collection('renders').getFullList<RendersResponse>({
					sort: '-created'
				})
			}
		};
}
```


**streaming on  +page.svelte**

```
...

let renders: RendersResponse[] | undefined = $state();

$effect(() => {
    data.streaming?.renders.then((result) => (renders = result));
})

...

{#each renders as render (render.id)}
....
{/each}


...
```
